### ISSUE TITLE
segmentation fault in clGetDeviceIDs on arm64(Jetson AGX)

### ISSUE THREAD

**Issue by buni-rock** [pocl/pocl#1206]
Hi,

after successfully compiled the pocl on Jetson AGX platform using LLVM 15, I have encountered a segmentation fault while calling the **clGetDeviceIDs** function. The compilation was successfully only after applying the patch mentioned in https://github.com/pocl/pocl/issues/1196

Trying to debugg, the issue, I could trace it till here:
![image](https://user-images.githubusercontent.com/30657575/231800571-90ecf782-62bf-4978-8e6a-ac6b821c9c6a.png)

**Comment by pjaaskel** [comment]
Hi. Generally the first step in debugging segfaults is usually to get a backtrace with the debug build. Does that show something more useful?

**Comment by franz** [comment]
This seems like a different issue than #1196. Do you have a backtrace ? Screenshots aren't too useful, but i see one problem: `dev->dispatch` is `0x1` - not a valid pointer. Are you using an ICD (libOpenCL.so) ? Is it Nvidia's ICD ? i would not bet that Nvidia's ICD works with anything but Nvidia's own OpenCL runtime. Can you try with opensource OCL-ICD or Khronos' ICD, or build PoCL without ICD (`-DENABLE_ICD=0`) ?

**Comment by buni-rock** [comment]
Hi again,

ICD:
![image](https://github.com/pocl/pocl/assets/30657575/9e92b198-6f28-423c-a7ed-93562f5d374a)

As stated in the documentation for OpenCL 3.0, ocl-icd (2.3.x) is necessary (which seems to no be available on my system)

Furthermore, clinfo reports the following
![image](https://github.com/pocl/pocl/assets/30657575/be73dcfa-0c37-4229-b5f4-4d04b41825e7)

The crash I have reported earlier happens under the `HOST_DEVICE_CL_VERSION_MAJOR >= 3` which would make sense according to the message reported by clinfo.

I have checked where the HOST_DEVICE_CL_VERSION_MAJOR is set, and it looks like it is somehow hardcoded to the compiler version: https://github.com/pocl/pocl/blob/release_3_1/CMakeLists.txt#L1187

On my side, I have installed LLVM 14 and 15; I am wondering if is possible to compile pocl using these compilers and not using OpenCL 3.0

I have built pocl with (-DENABLE_ICD=0) and afterwards clinfo outputs` Number of platforms                               0`

**Comment by heweiwill** [comment]
Hi @buni-rock ,

Are you able to run POCL (opencl version 3.0) on Jetson AGX ?   I am considering to purchase an Jetson Orin nano to running POCL, but not sure if it is viable, appreciate if you can advise

**Comment by isuruf** [comment]
If you are okay with using binaries, you can try the instructions at https://github.com/pocl/pocl#pocl-with-cuda-driver

**Comment by heweiwill** [comment]
Have this binary been verified work on any Jetson devices?

**Comment by isuruf** [comment]
I've tried it on an aarch64 server, but not a Jetson in particular. If it does not work there, I'll be happy to fix it.

**Comment by buni-rock** [comment]
> Hi @buni-rock ,
> 
> Are you able to run POCL (opencl version 3.0) on Jetson AGX ? I am considering to purchase an Jetson Orin nano to running POCL, but not sure if it is viable, appreciate if you can advise

Hi,

on our side we had 2 Jetsons, the one I tried to use failed to run POCL out of the box. The other one, which my colleague used, could run POCL without much effort. On the second device we could build POCL from sources and install it successfully. In principle, you can run POCL on Jetson, but the questions would be: how much effort you have to put in in order to achieve it?

PS: regarding this issue, I am afraid I can't proceed further due to lack of time.

**Comment by heweiwill** [comment]
@isuruf 
I followed the instruciton in https://github.com/pocl/pocl#pocl-with-cuda-driver, installed the pocl-cuda via mamba. But the clinfo still shows nothing.  Do you have idea how could i utililize the pocl installed?
BTW, i am still using an old jetson nano, not orin nano.

**Comment by heweiwill** [comment]
> > Hi @buni-rock ,
> > Are you able to run POCL (opencl version 3.0) on Jetson AGX ? I am considering to purchase an Jetson Orin nano to running POCL, but not sure if it is viable, appreciate if you can advise
> 
> Hi,
> 
> on our side we had 2 Jetsons, the one I tried to use failed to run POCL out of the box. The other one, which my colleague used, could run POCL without much effort. On the second device we could build POCL from sources and install it successfully. In principle, you can run POCL on Jetson, but the questions would be: how much effort you have to put in in order to achieve it?
> 
> PS: regarding this issue, I am afraid I can't proceed further due to lack of time.

what are the differences between these two Jetson?  why one works but the other one fails?

**Comment by isuruf** [comment]
>  I followed the instruciton in https://github.com/pocl/pocl#pocl-with-cuda-driver, installed the pocl-cuda via mamba. But the clinfo still shows nothing. Do you have idea how could i utililize the pocl installed?

You will see a `pocl.icd` file in `~/mambaforge/etc/OpenCL/vendors/pocl.icd`. Copy that to `/etc/OpenCL/vendors/pocl.icd`.
If you have the ICD loader from Khronos or ocl-dev, then you can set `OCL_ICD_VENDORS=~/mambaforge/etc/OpenCL/vendors/pocl.icd` instead of copying.

**Comment by heweiwill** [comment]
> > I followed the instruciton in https://github.com/pocl/pocl#pocl-with-cuda-driver, installed the pocl-cuda via mamba. But the clinfo still shows nothing. Do you have idea how could i utililize the pocl installed?
> 
> You will see a `pocl.icd` file in `~/mambaforge/etc/OpenCL/vendors/pocl.icd`. Copy that to `/etc/OpenCL/vendors/pocl.icd`. If you have the ICD loader from Khronos or ocl-dev, then you can set `OCL_ICD_VENDORS=~/mambaforge/etc/OpenCL/vendors/pocl.icd` instead of copying.

Hi @isuruf 
yes, i got the clinfo output, but some error from that , do you have any idea of that?
![Screenshot from 2023-07-06 14-45-45](https://github.com/pocl/pocl/assets/3840170/f40bdbf2-664a-49b6-864a-bf33c8eb2c5b)

**Comment by heweiwill** [comment]
maybe i should raise a new issue?

**Comment by isuruf** [comment]
Can you try with `export POCL_DEVICES=cuda`?

**Comment by heweiwill** [comment]
> Can you try with `export POCL_DEVICES=cuda`?

still report same error
![Screenshot from 2023-07-07 09-48-31](https://github.com/pocl/pocl/assets/3840170/8f772d4b-b6dd-4756-90e5-1ccf8d79b10e)

**Comment by heweiwill** [comment]
@isuruf  Do you have any idea on this -33 error code ?

**Comment by jansol** [comment]
According to [this handy table](https://streamhpc.com/blog/2013-04-28/opencl-error-codes/) -33 is `CL_INVALID_DEVICE` which would point towards something going wrong in device initialization. Try running with `POCL_DEBUG=error,cuda` and see if that spits out anything useful.

**Comment by heweiwill** [comment]
> According to [this handy table](https://streamhpc.com/blog/2013-04-28/opencl-error-codes/) -33 is `CL_INVALID_DEVICE` which would point towards something going wrong in device initialization. Try running with `POCL_DEBUG=error,cuda` and see if that spits out anything useful.

Hi @jansol 
Thanks very much for your reply. I add the debug environment variable, and got below log:
![Screenshot from 2023-07-10 11-08-55](https://github.com/pocl/pocl/assets/3840170/baaef05e-c8c1-4ba8-b175-d54a9a5160d7)

It seems related to the hardware architecture, i am using the Jetson nano devkit, the spec could be found at https://developer.nvidia.com/embedded/jetson-nano-developer-kit

**Comment by isuruf** [comment]
Can you try setting `export POCL_CUDA_GPU_ARCH=sm_53`?

**Comment by heweiwill** [comment]
@isuruf 
Still failed at initialization, log info below:
![Screenshot from 2023-07-11 09-55-53](https://github.com/pocl/pocl/assets/3840170/fff2f9b2-e7d8-447f-a1c4-944231f9967c)

**Comment by heweiwill** [comment]
I tried sm_50, sm_52 for Maxwell architecture as well, all failed, report same error

**Comment by onfe** [comment]
A good sanity check is to install `nvidia-cuda-toolkit` and see if you can pickup the GPU with `nvidia-smi`. I've found the Nvidia drivers for `aarch64` are a bit picky. Not that this will necessarily solve your issue, but I ended up re-installing my Nvidia drivers on an AVA devkit.

I'm looking forward to seeing your progress on this issue.

**Comment by isuruf** [comment]
I guess the following works?
```
#include <stdio.h>
#include <cuda.h>
#include <cuda_runtime.h>

#define CUDA_CHECK(ans) { gpuAssert((ans), __FILE__, __LINE__); }
void gpuAssert(cudaError_t code, const char *file, int line)
{
   if (code != cudaSuccess) 
   {
      fprintf(stderr,"GPUassert: %s %s %d\n", cudaGetErrorString(code), file, line);
      exit(code);
   }
}

int main()
{
  CUDA_CHECK(cuInit(0));
  CUdevice device;
  CUDA_CHECK(cuDeviceGet(&device, 0));
  CUcontext ctx;
  CUDA_CHECK(cuCtxCreate(&ctx, 0, device));
  CUDA_CHECK(cuCtxCreate(&ctx, 0, device));
  int sm_maj, sm_min;
  CUDA_CHECK(cuDeviceGetAttribute(&sm_maj, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, device));
  CUDA_CHECK(cuDeviceGetAttribute(&sm_min, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR, device));
  printf("sm_%d%d\n", sm_maj, sm_min);
  CUDA_CHECK(cuCtxDestroy(ctx));
  return 0;
}
```

**Comment by heweiwill** [comment]
> A good sanity check is to install `nvidia-cuda-toolkit` and see if you can pickup the GPU with `nvidia-smi`. I've found the Nvidia drivers for `aarch64` are a bit picky. Not that this will necessarily solve your issue, but I ended up re-installing my Nvidia drivers on an AVA devkit.
> 
> I'm looking forward to seeing your progress on this issue.

Thanks Edward for the suggestion.  It's weird i cannot install nvidia toolkit at my jetson nano. I will try to figure it out later.

**Comment by heweiwill** [comment]
The NV provided device query code output below information:
![Screenshot from 2023-07-12 10-07-26](https://github.com/pocl/pocl/assets/3840170/9a9ce886-3bb3-48d9-81b2-1047fac6349d)

**Comment by heweiwill** [comment]
From the log of this example code, the Jetson nano compute capability truely is sm53:
![Screenshot from 2023-07-12 11-16-00](https://github.com/pocl/pocl/assets/3840170/303ff4a3-8982-4e84-a1c7-bd2ee48893d6)

**Comment by heweiwill** [comment]
Hi @isuruf 
any other ideas?

**Comment by isuruf** [comment]
Can you try https://github.com/pocl/pocl/issues/1206#issuecomment-1631029796 ?

**Comment by heweiwill** [comment]
> Can you try [#1206 (comment)](https://github.com/pocl/pocl/issues/1206#issuecomment-1631029796) ?

Do you mean the code below you mentioned?  I tried the example code SimpleTextureDrv, which include the same function you provided, and it does able to read the SM major and minor value, and the value is 53, same as previous we manually set as the environment variable.


#include <stdio.h>
#include <cuda.h>
#include <cuda_runtime.h>

#define CUDA_CHECK(ans) { gpuAssert((ans), __FILE__, __LINE__); }
void gpuAssert(cudaError_t code, const char *file, int line)
{
   if (code != cudaSuccess) 
   {
      fprintf(stderr,"GPUassert: %s %s %d\n", cudaGetErrorString(code), file, line);
      exit(code);
   }
}

int main()
{
  CUDA_CHECK(cuInit(0));
  CUdevice device;
  CUDA_CHECK(cuDeviceGet(&device, 0));
  CUcontext ctx;
  CUDA_CHECK(cuCtxCreate(&ctx, 0, device));
  CUDA_CHECK(cuCtxCreate(&ctx, 0, device));
  int sm_maj, sm_min;
  CUDA_CHECK(cuDeviceGetAttribute(&sm_maj, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, device));
  CUDA_CHECK(cuDeviceGetAttribute(&sm_min, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR, device));
  printf("sm_%d%d\n", sm_maj, sm_min);
  CUDA_CHECK(cuCtxDestroy(ctx));
  return 0;
}

**Comment by isuruf** [comment]
Ah, you have CUDA 10.2, can you update to CUDA 11.1+?

**Comment by isuruf** [comment]
I updated the conda package with https://github.com/pocl/pocl/pull/1277. Can you try `conda update pocl-cuda` to get the latest package?

**Comment by heweiwill** [comment]
> I updated the conda package with #1277. Can you try `conda update pocl-cuda` to get the latest package?

Does it mean CUDA10.2 also can work?   

As i tried to upgrade the CUDA10.2 to CUDA11.3, but failed, the jetson cannot  boot up after upgrade,  it turns out that the CUDA on jetson nano is highly coupled with low level driver and not able to simply upgrade CUDA, but need NV to provide an whole new image for jetson nano to support newer version CUDA.  I learned from NV developer that after CUDA11.8, then the jetson able to support upgrade CUDA solely.

**Comment by isuruf** [comment]
> Does it mean CUDA10.2 also can work?

Yes

**Comment by heweiwill** [comment]
@isuruf
Many thanks for your help!  I am able to get clinfo output at my jetson nano :)
[jetson-nano_clinfo.txt](https://github.com/pocl/pocl/files/12180960/jetson-nano_clinfo.txt)

**Comment by heweiwill** [comment]
@isuruf 
I am wondering does POCL have any function tests for cuda?  I have an opencl application runs well on ubuntu x86 with cuda, but just hang when running on jetson nano. It would be good to have a test that POCL works well with cuda10.2 version.

**Comment by isuruf** [comment]
Does it work with CPU driver on nano? Does it work with PoCL CUDA on x86?

We do have tests, but PoCL on CUDA does not pass all of them and we do not run tests on aarch64 architectures due to the lack of hardware.

**Comment by pjaaskel** [comment]
run_cuda_tests.sh should run only the tests that are expected to pass on cuda.
